### PR TITLE
[custom_op] triton_op API V0

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130637

This is the initial version of an API to create custom operators whose
implementations are backed by triton kernels. While user-defined triton
kernels work out-of-the-box with triton kernels, you may wish to
construct a custom operator if you need to compose with other PyTorch
subsystems, like Tensor subclasses or vmap.

I'm hoping to get design feedback on this and ship it so that we can
begin experimenting with customers.

Test Plan:
- new tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_library/triton.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        result.register_torch_dispatch(FunctionalTensorMode, functional_decomp)
+        return result
+
+    if fn is None:
+        return dec
+    else:
+        return dec(fn)
+
+
+def capture_triton(triton_kernel: Callable, /) -> Callable:
```
Comment (zou3519) [inline torch/_library/triton.py:123]: This was moved from torch/_higher_order_ops/triton_kernel_wrap.py

File: torch/_library/triton.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+    Args:
+        name (str): A name for the custom op that looks like "{namespace}::{name}",
+            e.g. "mylib::my_linear". The name is used as the op's stable identifier
+            in PyTorch subsystems (e.g. torch.export, FX graphs).
+            To avoid name collisions, please use your project name as the namespace;
+            e.g. all custom ops in pytorch/fbgemm use "fbgemm" as the namespace.
+        mutates_args (Iterable[str] or "unknown"): The names of args that the function mutates.
+            This MUST be accurate, otherwise, the behavior is undefined. If "unknown",
+            it pessimistically assumes that all inputs to the operator are being mutated.
```
Comment (zou3519) [inline torch/_library/triton.py:35]: This is required right now, but in theory we can infer it or assume all functional and error out if we detect that isn't the case. I can handle that in a follow-up.

File: torch/_library/triton.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        >>>
+        >>> x = torch.randn(3, device="cuda")
+        >>> y = torch.randn(3, device="cuda")
+        >>>
+        >>> z = f(x, y)
+        >>> assert torch.allclose(z, x + y)
+
+    """
+
+    def dec(fn: Callable) -> Callable:
```
Comment (zou3519) [inline torch/_library/triton.py:92]: The interesting thing about `@triton_op` is that there is nothing triton-specific about it at the moment. This seems like more of a "how do we create a custom op that decomposes right before Inductor" type of deal. But I think triton kernels are big enough that they warrant their own high-level API.

File: torch/_higher_order_ops/triton_kernel_wrap.py [hunk 4]
```diff
@@ -1071,17 +998,13 @@ def __getitem__(self, *args):
         return tracing_triton_hopifier_singleton.call_getitem(self, args)
 
     def run(self, *args, **kwargs):
-        import torch._dynamo
-
-        if not is_fx_tracing() or torch._dynamo.is_compiling():
+        if False:
```
Comment (albanD) [inline torch/_higher_order_ops/triton_kernel_wrap.py:1001]: I guess these will be fixed?

File: torch/_library/triton.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        # - With post-dispatch torch.export, this means that there will
+        #   be a call(s) to the triton_kernel_wrapper_functional HOP in the
+        #   graph (that we have yet to figure out how to serialize).
+        def functional_decomp(  # type: ignore[no-untyped-def]
+            mode, _, types, args, kwargs
+        ):
+            with mode:
+                return fn(*args, **kwargs)
+
+        result.register_torch_dispatch(FunctionalTensorMode, functional_decomp)
```
Comment (albanD) [inline torch/_library/triton.py:114]: Very cool!

**Review by albanD — APPROVED** [review]
Looks great!

**Pull request closed without merging**